src/p2p/net_node.inl

```diff
@@ -1704,4 +1749,30 @@ namespace nodetool
 
     return count > max_connections;
   }
+
+  template<class t_payload_net_handler>
+  bool node_server<t_payload_net_handler>::gray_peerlist_housekeeping()
+  {
+    peerlist_entry pe = AUTO_VAL_INIT(pe);
+
+    if (!m_peerlist.get_gray_peer_random(pe)) {
+        return false;
+    }
+
+    bool success = check_connection_and_handshake_with_peer(pe.adr, pe.last_seen);
+
+    if (!success) {
+      m_peerlist.remove_from_peer_gray(pe);
```

Comment:
I think most unreachable peers in the gray list will be:
 - nodes on dynamic IPs - they'll be reachable again, but maybe in a long time
 - nodes that come and go - they'll be reachable at some point in the near future

The second ones should be kept I think.

I would evict only if last seen is more than, say, a month.